### PR TITLE
Modularized MatrixFreeTools::compute_diagonal()

### PR DESCRIPTION
**Pull request by peterrum**
~Depends on #11803.~

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3992,9 +3992,13 @@ inline const MatrixFree<dim, Number, VectorizedArrayType> &
 FEEvaluationBaseData<dim, Number, is_face, VectorizedArrayType>::
   get_matrix_free() const
 {
+<<<<<<< HEAD
```
Comment (Rombur) [inline include/deal.II/matrix_free/fe_evaluation.h:3995]: You forgot to fix this merge conflict.

Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: You are right. I have fixed it!

**Comment by peterrum** [comment]
The logic of `MatrixFreeTools::compute_diagonal()` has been moved to a new helper class. The usage is like this:
```cpp
internal::ComputeDiagonalHelper<...> helper(phi);

for (unsigned int cell = range.first; cell < range.second; ++cell)
  {
    helper.reinit(cell);

    for (unsigned int i = 0; i < phi.dofs_per_cell; ++i)
      {
        helper.prepare_basis_vector(i);
        local_vmult(phi);
        helper.submit();
      }

    helper.distribute_local_to_global(diagonal_global);
  }
```
and users can add additional operations at any place.

**Review by kronbichler — APPROVED** [review]
I assume this is now already tested by the existing `compute_diagonal` tests.

**Comment by peterrum** [comment]
> I assume this is now already tested by the existing compute_diagonal tests.

Yes.

**Pull request merged**